src/core/data/store.cc

```diff
@@ -123,12 +133,27 @@ FutureWrapper::FutureWrapper(
   assert(field_size > 0);
 #endif
   if (!read_only) {
-    auto mem_kind = find_memory_kind_for_executing_processor();
 #ifdef DEBUG_LEGATE
     assert(!initialize || future_.get_untyped_size() == field_size);
 #endif
-    auto p_init_value = initialize ? future_.get_buffer(mem_kind) : nullptr;
-    buffer_           = UntypedDeferredValue(field_size, mem_kind, p_init_value);
+    auto proc     = Processor::get_executing_processor();
+    auto mem_kind = proc.kind() == Processor::Kind::TOC_PROC ? Memory::Kind::GPU_FB_MEM
+                                                             : Memory::Kind::SYSTEM_MEM;
+    if (initialize) {
+      auto p_init_value = future_.get_buffer(mem_kind);
+#ifdef LEGATE_USE_CUDA
+      if (mem_kind == Memory::Kind::GPU_FB_MEM) {
+        // TODO: This should be done by Legion
+        buffer_ = UntypedDeferredValue(field_size, mem_kind);
+        AccessorWO<int8_t, 1> acc(buffer_, field_size, false);
+        auto stream = cuda::StreamPool::get_stream_pool().get_stream();
+        CHECK_CUDA(
+          cudaMemcpyAsync(acc.ptr(0), p_init_value, field_size, cudaMemcpyDeviceToDevice, stream));
```

Comment:
Do we also need to deviceSync in this case, in case the task performs cuda operations on a different stream?

Reply:
https://github.com/nv-legate/legate.core/pull/353/commits/7f07fa116925cb78585f3ae4b161d64c8b4bb148